Employee standing report

lomamech opened the pull request:
- Implement global situation of all employees in different mode

closes #6123

lomamech committed:
Update form for Employee standing

lomamech committed:
Refactor (Employees Standing report)
- Implement global situation of all employees in different mode

closes #6123

jmcameron reviewed (commented):
Looks good.  I tried it and it seems to work well.  However, I have cosmetic and translation suggestions.
- I suggest NOT underlining the report title.   None of our other reports titles use underlining.  I think h3 + bold is enough.
- When selecting all employees, I see the term "Synthetic".  This is a little obscure in English.  I think what you are thinking of is a 'synthesis' or summary.  If that is correct, I suggest  simply using the term 'Summary' instead of 'Synthetic'.   Also, the term 'Aggregate' is a bit odd to my eye.  Not sure what to suggest, but the terms 'Normal Report', 'Full Report', 'Report for All Employees', etc.  I like the last option 'Detailed report'.   Finally, I suggest moving the "Synthetic" option to the last one in the list since the other two are more closely related.
- I suggest rewording the label for the last option from "View extract of the employee's standing by date range" to something simpler such as "Limit to time interval".   I know what you mean by "Extract" but I'm not sure everyone will.

lomamech committed:
improvement(Employee Standing)
- Improve, clean and sanitaze code

jmcameron reviewed (commented):
I like your changes.  One very minor cosmetic thing:  You have several 'hr' separators in the  client-side page.  I don't think they are necessary and add extra unneeded white space.

I noticed one thing that may be an issue.  What is the purpose of the "Limit to time interval" option?  If I select one employee and select various time intervals it does not seem to make a difference.  But it does make a difference for normal/summary reports.  And why would I want to use this option anyway?  Don't we always want to know the current standing?   And why chose and interval?   It seems it would make more sense to me to want to know the employee standing at some specific time in the past.   My understanding of what you're trying to do is limited so I encourage @jniles to review this PR.

lomamech committed:
feature(Employee Standing)
- Adding the selection of currency
- Adding in action Menu in the employee registry

closes #6126

jmcameron commented on client/src/modules/reports/generate/employee_standing/employee_standing.js:
```diff
@@ -12,21 +12,28 @@ EmployeeStandingController.$inject = [
 
   checkCachedConfiguration();
 
+  vm.reportDetails.currency_id = Session.enterprise.currency_id;
```
Note:  Doing it this way forces the form to use the enterprise currency each time you enter the page.  The cache is ignored.  Look at the checkCachedConfiguration() function in client/src/modules/reports/generate/avg_med_costs_per_patient/avg_med_costs_per_patient.js for an example.

jniles replied:
Yes, the best way to do this would be to check if the `currency_id` is defined, and only set it if it isn't defined.

Like this:

```js
vm.reportDetails.currency_id = vm.reportDetails?.currency_id || Session.enterprise.currency_id;
```

See [Optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) for more information on how this works.

lomamech replied:
Nice

jmcameron commented on server/controllers/finance/reports/financial.employee.handlebars:
```diff
@@ -14,13 +14,13 @@
   <h4 class="text-uppercase" style="background-color: #efefef;">
     <b>
-      {{translate balanceCreditorText}} = {{debcred ./creditorAggregates.balance metadata.enterprise.currency_id}}
+      {{translate balanceCreditorText}} = {{debcred (multiply ./creditorAggregates.balance exchangeRate) currencyId}}
```
Nice!

jmcameron commented on server/controllers/finance/reports/financial.employee.js:
```diff
@@ -49,8 +58,10 @@ async function build(req, res, next) {
-
     const data = {};
+    data.exchangeRate = exchange.rate || 1;
+    data.currencyId = options.currency_id;
```
See my previous note.

jmcameron reviewed (commented):
Thanks for adding the multi-currency support.  I made one suggestion in the code about caching the currency used in the report.

As far as I know, once you address the currency caching and rebase, it is ready to go.  But I do not understand how the time limits work (see my previous reviews above).  So I still suggest getting @jniles to review this before merging.

jniles commented on client/src/modules/reports/generate/employee_standing/employee_standing.html:
```diff
@@ -22,22 +22,76 @@ <h3 class="text-capitalize" translate>REPORT.EMPLOYEE_STANDING.TITLE</h3>
+              <label class="radio-inline">
+                <input
+                  type="radio"
+                  name="modeRepport"
```
Can you make this name `modeReport`?  `Report` doesn't have two `pp`s in English 😄

jniles commented on client/src/modules/reports/generate/employee_standing/employee_standing.html:
```diff
@@ -22,22 +22,76 @@ <h3 class="text-capitalize" translate>REPORT.EMPLOYEE_STANDING.TITLE</h3>
+                  type="radio"
+                  name="modeRepport"
+                  value="normal"
+                  ng-model="ReportConfigCtrl.reportDetails.modeRepport"
```
Same with the `ngModel`

jniles commented on client/src/modules/reports/generate/employee_standing/employee_standing.html:
```diff
@@ -22,22 +22,76 @@ <h3 class="text-capitalize" translate>REPORT.EMPLOYEE_STANDING.TITLE</h3>
+                <input
+                  type="radio"
+                  name="modeRepport"
+                  value="detailled"
```
`detailed` has only one `l` in English 😄

jniles commented on client/src/modules/reports/generate/employee_standing/employee_standing.js:
```diff
@@ -41,12 +48,24 @@ function EmployeeStandingController($state, $sce, Notify, BaseReportService, App
+    if (vm.reportDetails.allEmployee) {
+      delete vm.reportDetails.employee_uuid;
+      delete vm.reportDetails.includeMedicalCare;
+      reportUrl = '/reports/finance/employees_standing';
```
These two URLs are really close together!  I thought you had a typo at first.  Is it possible to either

1. Make them really different, so we don't accidentally correct this as a typo OR
2. Make the switch between all employees just a query string parameter and use the same base URL?

jniles commented on server/controllers/finance/reports/financial.employees.js:
```diff
@@ -0,0 +1,206 @@
+  // set up the report with report manager
+  try {
+    report = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
```
Can you wrap this entire function block in a `try/catch` statement?  I don't see the need to have several different `try/catch` blocks, and if something goes wrong on the `await Exchange.getExhangeRate()` (like if MySQL crashes), you won't catch it.

jniles commented on server/controllers/finance/reports/financial.employees.js:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    const [financialData] = await Promise.all([
+      db.exec(sql),
```
You don't need `Promise.all()` for a single query.

jniles reviewed (commented):
@lomamech this looks pretty good!

If you show values in multiple currencies, you need to have a date and exchange rate shown on the report, as well as a warning when the user does not choose the enterprise exchange rate:

![image](https://user-images.githubusercontent.com/896472/146509665-b752f1b3-6327-4a10-a4f6-84313b015ca8.png)

Otherwise, I agree with @jmcameron commentary.

lomamech committed:
improvement(All Employee Standing report)
- Cache the currency used
- Using optional chaining
- Display warning when the user does not choose the enterprise exchage
  rate
- Update label using report instead repport and detailed instead
  detailled
- Update url for Report of All Employees
- Use only one try/Catch
- Improve the using of function Exchange.getExchangeRate() use the
  dateTo for getting report

lomamech commented on client/src/modules/reports/generate/employee_standing/employee_standing.js:
```diff
@@ -12,21 +12,29 @@ EmployeeStandingController.$inject = [
 
   checkCachedConfiguration();
 
+  vm.reportDetails.currency_id = vm.reportDetails?.currency_id || Session.enterprise.currency_id;
```
@jniles  Optional chaining is not **Recognized**

jniles replied:
Yeah, hound is probably wrong here.  Let me try updating the version of ESLint used by hound.
